pywhy_stats/categorical.py

```diff
@@ -0,0 +1,293 @@
+) -> PValueResult:
+    """Compute the Cressie-Read power divergence statistic.
+
+    The null hypothesis for the test is X is independent of Y given Z. A lot of the
+    frequency comparison based statistics (eg. chi-square, G-test etc) belong to
+    power divergence family, and are special cases of this test.
+
+    Parameters
+    ----------
+    X: ArrayLike of shape (n_samples,)
+        The first node variable.
+    Y : ArrayLike of shape (n_samples,)
+        The second node variable.
+    Z : ArrayLike of shape (n_samples, n_variables)
+        The conditioning set.
+    lambda_: float or string
+        The lambda parameter for the power_divergence statistic. Some values of
+        lambda_ results in other well known tests:
+            "pearson"             1          "Chi-squared test"
+            "log-likelihood"      0          "G-test or log-likelihood"
+            "freeman-tukey"     -1/2        "freeman-tukey Statistic"
+            "mod-log-likelihood"  -1         "Modified Log-likelihood"
+            "neyman"              -2         "Neyman's statistic"
+            "cressie-read"        2/3        "The value recommended in the paper
+                                             :footcite:`cressieread1984`"
+
+    Returns
+    -------
+    CI Test Results: tuple
+        Returns a tuple (chi, p_value, dof). `chi` is the
+        chi-squared test statistic. The `p_value` for the test, i.e. the
+        probability of observing the computed chi-square statistic (or an even
+        higher value), given the null hypothesis that X \u27C2 Y | Zs is True.
+        If boolean = True, returns True if the p_value of the test is greater
+        than `significance_level` else returns False.
+
+    See Also
+    --------
+    scipy.stats.power_divergence
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if X.ndim != 1 or Y.ndim != 1:
+        raise ValueError("X and Y should be 1-D arrays.")
+
+    # ensure we use numpy arrays
+    X = np.asarray(X)
+    Y = np.asarray(Y)
+
+    # Check if all elements are integers
+    if not np.issubdtype(X.dtype, np.integer):
+        le = LabelEncoder()
+        X = le.fit_transform(X)
+        # warn("Converting X array to categorical array using scikit-learn's LabelEncoder.")
+
+    # Check if all elements are integers
+    if not np.issubdtype(Y.dtype, np.integer):
+        le = LabelEncoder()
+        Y = le.fit_transform(Y)
+        # warn("Converting Y array to categorical array using scikit-learn's LabelEncoder.")
+
+    for name, arr in zip(["X", "Y"], [X, Y]):
+        # Check if the number of unique values is reasonably small
+        unique_values = np.unique(arr)
+        num_unique_values = len(unique_values)
+        if num_unique_values > 64:  # Adjust the threshold as needed
+            raise RuntimeError(
+                f"There are {num_unique_values} unique categories for {name}. "
+                f"This is likely an error."
+            )
```

Comment:
Where does this 64 come from? Just a random choice?

Reply:
Yes random choice. For a category to have over 64 possible states is a lot. Honestly I think it can be even lower because if we're talking about discrete categorical data, I would be surprised if there is more than 10 states.

If it's an issue w/ users, we can revisit this issue? I can add a `# XXX:` inline comment.